torch/csrc/distributed/c10d/init.cpp

```diff
@@ -62,7 +62,8 @@ PyObject* c10d_init(PyObject* _unused) {
           [](::c10d::Reducer& reducer, const torch::autograd::Variable& output)
               -> void { reducer.prepare_for_backward({output}); },
           py::call_guard<py::gil_scoped_release>())
-      .def("get_backward_stats", &::c10d::Reducer::get_backward_stats);
+      .def("get_backward_stats", &::c10d::Reducer::get_backward_stats)
+      .def("remove_all_hooks", &::c10d::Reducer::remove_all_hooks);
```

Comment:
The Python binding can be removed I think.

Reply:
Kutta asked for exposing this API, just in case Python GC does not deterministically kick in. @kuttas do you still need this?

Reply:
@pietern Even if we call `__del__` on the `DistributedDataParallel` object I am not totally sure that the C++ destructor (Reducer::~Reducer) will get called immediately or lazily (GC time). If this is 100% deterministic, then we don't need this call. Otherwise, we need to implement `__del__` and explicitly clean up hooks in DDP. I searched pybind documentation and can't figure out what the expected behavior is if you don't have referential cycles - is ref count update immediate in python and does that immediately trigger the C++ destructor of the py-bound class?

Reply:
It doesn't matter if it is lazily deleted. The hooks may still be called, but only as long as the reducer object is still alive (per the destructor). If the `prepare_for_backward` function isn't called by the DDP class that wraps the reducer, all the hooks will be nops.

Regarding referential cycles -- IIRC this is where the GC comes in. Objects can be destructed immediately if their refcount drops to zero and then there is a separate GC pass for the cycles.